st_sampler.py

```diff
@@ -0,0 +1,168 @@
+import random
+
+# given a walk as a list of vertices
+# returns a newwalk which has no loops
+# uses -1 as a invalid value
+def eraseLoops(walk):
+    result = [-1] * (len(walk) + 1)
+    ptr = 0
+    seen = {}
+    for v in walk:
+        if v in seen:
+            newptr = seen[v] # erase everything until this point
+            for i in range(newptr+1, ptr): # remove these node from seen
+                del seen[result[i]]
+            ptr = newptr + 1
+
+        else:
+            result[ptr] = v # add to result
+            seen[v] = ptr # record this location
+            ptr += 1
+        result[ptr] = -1 # do this to mark the next location
+
+    # copy the result into an array of correct size
+    res = []
+    for x in result:
+        if x == -1:
+            return res
+        else:
+            res.append(x)
+        
+    return res
+
+# class to sample random spanning trees
+class STSampler:
+    def __init__(self, g):  
+        self.graph = g # graph is adjList stored in a dict
+
+    # choose a neighbor of x in self.graph
+    def chooseNeighbor(self, x):
+        if x in self.graph:
+            return random.choice(self.graph[x])
+        else:
+            print('This node is not in the graph!')
+            return -1
+        
+    # starting from u, do a random walk until we meet the tree
+    # assumes that u is not already part of the tree
+    def findWalk(self, u, tree):
+        walk = [u]
+        while walk[-1] not in tree: # while not yet reached tree
+            cur = walk[-1]
+            nxt = self.chooseNeighbor(cur)
+            walk.append(nxt)
+        return walk
+        
+    # adds the found path to the tree
+    def addPath(self, tree, path):
+        for i in range(0, len(path) - 1):
+            tree[path[i]] = [] # add all the vertices
+        for i in range(0, len(path) - 1): # add the edges
+            u = path[i]
+            v = path[i+1]
+            tree[u].append(v) # can be improved to check for duplicates
```

Comment:
Would there be duplicates? If there is a duplicate, then either 1. the path contains a loop, or 2. there is more than 1 vertex in the path that is already in the tree

Reply:
If we only add paths that have loops erased, then will be fine, but i put that comment there so that if got more time maybe can make it more robust 😂